src/core/Table/hooks/useColumnDragAndDrop.tsx

```diff
@@ -0,0 +1,140 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import React from 'react';
+import {
+  HeaderGroup,
+  Hooks,
+  IdType,
+  makePropGetter,
+  TableInstance,
+  TableKeyedProps,
+  useGetLatest,
+} from 'react-table';
+
+export const useColumnDragAndDrop = <T extends Record<string, unknown>>(
+  isEnabled: boolean,
+) => (hooks: Hooks<T>) => {
+  hooks.getDragAndDropProps = [defaultGetDragAndDropProps(isEnabled)];
+  hooks.useInstance.push(useInstance);
+};
+
+const defaultGetDragAndDropProps = <T extends Record<string, unknown>>(
+  isEnabled: boolean,
+) => (
+  props: TableKeyedProps,
+  {
+    instance,
+    header,
+  }: {
+    instance: TableInstance;
+    header: HeaderGroup;
+  },
+) => {
+  if (!isEnabled || header.disableDragging) {
+    return props;
+  }
+
+  const onDragStart = (event: React.DragEvent<HTMLDivElement>) =>
+    event.dataTransfer.setData('text', header.id);
+
+  const setOnDragColumnStyle = (
+    event: React.DragEvent<HTMLDivElement>,
+    position?: 'left' | 'right',
+  ) => {
+    const columnElement = event.currentTarget as HTMLElement;
+    columnElement.classList.remove('iui-reorder-column-right');
+    columnElement.classList.remove('iui-reorder-column-left');
+    if (position === 'left') {
+      columnElement.classList.add('iui-reorder-column-left');
+    } else if (position === 'right') {
+      columnElement.classList.add('iui-reorder-column-right');
+    }
+  };
+
+  const reorderColumns = (
+    tableColumns: IdType<T>[],
+    srcIndex: number,
+    dstIndex: number,
+  ) => {
+    const newTableColumns = [...tableColumns];
+    const [removed] = newTableColumns.splice(srcIndex, 1);
+    newTableColumns.splice(
+      // When dragging column to the right, we need to decrease the index
+      // because we removed column before and indexes shifted to the left.
+      dstIndex > srcIndex ? dstIndex - 1 : dstIndex,
+      0,
+      removed,
+    );
+    return newTableColumns;
+  };
+
+  const onDragOver = (event: React.DragEvent<HTMLDivElement>) => {
+    event.preventDefault();
+    const columnElement = event.target as HTMLElement;
+    const middlePoint =
+      columnElement.offsetLeft + columnElement.offsetWidth / 2;
+    setOnDragColumnStyle(event, event.clientX > middlePoint ? 'right' : 'left');
```

Comment:
Maybe move this middle point logic into separate function? As you have 2 places which uses same.
`isDraggedToTheRight` or something like that.